api/src/utils/map/provider.ts

```diff
@@ -83,6 +90,14 @@ export const mapProviderToList = (provider: Provider, providerAttributeSchema: P
   } as ProviderList;
 };
 
+function getDistinctGpuModelsFromNodes(nodes: ProviderSnapshotNode[]) {
+  const gpuModels = nodes.flatMap((x) => x.gpus).map((x) => ({ vendor: x.vendor, model: x.name, ram: x.memorySize, interface: x.interface }));
+  const distinctGpuModels = gpuModels.filter(
+    (x, i, arr) => arr.findIndex((o) => x.vendor === o.vendor && x.model === o.model && x.ram === o.ram && x.interface === o.interface) === i
```

Comment:
Could you pls explain what this means?
I understand that `findIndex` matching same vendor, model, etc. Why should it also be of the same index?

Reply:
ok, I see, it's to remove duplicates

Reply:
It's really hard to understand this. I'd suggest if we could add a dedicated function for this.

```typescript
type Matcher<T> = (a: T, b: T) => boolean;

function createFilterUnique<T>(matcher: Matcher<T> = (a, b) => a === b): (value: T, index: number, array: T[]) => boolean {
  return (value, index, array) => {
    return array.findIndex((other) => matcher(value, other)) === index;
  };
}
```

in this case the code above would be:

```typescript
const distinctGpuModels = gpuModels.filter(
    createFilterUnique(x, o) => x.vendor === o.vendor && x.model === o.model && x.ram === o.ram && x.interface === o.interface))
);
```

and example [below](https://github.com/akash-network/cloudmos/pull/143/files#diff-11422e0f72849b5cbe75e9e428d718b2798c564d11f7e80f91f72f03a27030dcR29) would be

```typescript
.filter(createFilterUnique())
```

Reply:
can be tested [here](https://www.typescriptlang.org/play?#code/C4TwDgpgBAsghsAxgCwgJwDwBUB8UC8UAFHAFxRYA0UARuVgJQF40D2rANhHAHYDcAKAEAzAK49EwAJaseURGm7AIAMSkdlaAKo8pAR1ERsOIgFsEKdOXhJUmXAWJxqNJvjxwC+Qq-JEAbnAchvTUUjwAJhAAHuQ8oqY06NRwaGhwIPQA2gC6bizsXLxQAN4CUFCKwKJocgFBhmGRMSlpGfml5RWVENW1UKnpIAB0wuERAJLN0URErMB2Hea26PXBENTzi27eUOMxghUAvoJHQoiyAM7AUMLsjlkA5HCP1I80r1CPiJ-vvy85UbqTREBRKVTA9A6fSGIgMBjnK6cCDDDisADmRDurARAguPGut3YAEYHiUiaxyO9Uo8oEdqOTsVSaDS6QyKcy4AAvWn00ocr4stDE3mAsYaVZghAQiXaXQGCCzZy0DpwUb3bw+dU43H4y7I1EYrEkhhAA)

Reply:
It's to keep only the first occurence of each gpus since there is duplicates in the array. If the current index `i` matches `findIndex` that means it's the first occurence so it's kept, otherwise it is filtered out as duplicate.

Reply:
yea, I figured after asking the question :) check out a suggestion pls

Reply:
mb your other replies weren't showing when I answered. For sure that would be easier to understand, I'll add it! 👍